tools/witx/src/toplevel.rs

```diff
@@ -1,162 +1,144 @@
+    let mut validator = DocValidation::new();
+    let mut definitions = Vec::new();
+    let root = path.parent().unwrap_or(Path::new("."));
+
+    parse_file(
+        path.file_name().unwrap().as_ref(),
+        io,
+        root,
+        &mut validator,
+        &mut definitions,
+        &mut HashSet::new(),
+    )?;
+    Ok(Document::new(definitions, validator.entries))
 }
 
-fn resolve_uses(
-    toplevel: Vec<TopLevelSyntax>,
-    search_path: &Path,
-    used: &mut HashSet<PathBuf>,
-    witxio: &dyn WitxIo,
-) -> Result<Vec<DeclSyntax>, WitxError> {
-    let mut decls = Vec::new();
+fn parse_file(
+    path: &Path,
+    io: &dyn WitxIo,
+    root: &Path,
+    validator: &mut DocValidation,
+    definitions: &mut Vec<Definition>,
+    parsed: &mut HashSet<PathBuf>,
+) -> Result<(), WitxError> {
+    let path = io.canonicalize(&root.join(path))?;
+    if !parsed.insert(path.clone()) {
+        return Ok(());
+    }
+    let input = io.fgets(&path)?;
+
+    let adjust_err = |mut error: wast::Error| {
+        error.set_path(&path);
+        error.set_text(&input);
+        WitxError::Parse(error)
+    };
+    let buf = wast::parser::ParseBuffer::new(&input).map_err(adjust_err)?;
+    let doc = wast::parser::parse::<TopLevelDocument>(&buf).map_err(adjust_err)?;
 
-    for t in toplevel {
+    for t in doc.items {
         match t {
-            TopLevelSyntax::Decl(d) => decls.push(d),
+            TopLevelSyntax::Decl(d) => {
+                let def = validator
+                    .scope(&input, &path)
+                    .validate_decl(&d)
+                    .map_err(WitxError::Validation)?;
+                definitions.push(def);
+            }
             TopLevelSyntax::Use(u) => {
-                let abs_path = witxio.canonicalize(&search_path.join(u.name))?;
-                // Include the decls from a use declaration only once
-                // in a given toplevel. Same idea as #pragma once.
-                if !used.contains(&abs_path) {
-                    used.insert(abs_path.clone());
-
-                    let source_text = witxio.fgets(&abs_path)?;
-                    let inner_toplevels = parse_toplevel(&source_text, &abs_path)?;
-
-                    let inner_decls = resolve_uses(inner_toplevels, search_path, used, witxio)?;
-                    decls.extend(inner_decls)
-                }
+                parse_file(u.as_ref(), io, root, validator, definitions, parsed)?;
             }
         }
     }
 
-    Ok(decls)
+    Ok(())
 }
 
 #[cfg(test)]
 mod test {
     use super::*;
+    use crate::ast::*;
     use crate::io::MockFs;
-    use crate::parser::*;
-    use crate::Location;
 
     #[test]
     fn empty() {
-        assert_eq!(
-            parse_witx_with(&Path::new("/a"), &MockFs::new(&[("/a", ";; empty")])).expect("parse"),
-            Vec::new(),
-        );
+        parse_witx_with(&Path::new("/a"), &MockFs::new(&[("/a", ";; empty")])).expect("parse");
```

Comment:
Why did the `assert_eq!` part go away here? Should we still be able to assert that `part_witx_with` returns an empty result here?

Reply:
The interface of this function changed slightly to return a `Document` instead of returning a raw parsed syntax tree, so I ended up changing the tests too. I don't mind preserving the tests though, I could add a separte function to ensure everything is parsed as-is, but I figued it wouldn't add too much over the assertions this has currently.

Reply:
I wouldn't sweat these tests - I want to make changes soon that allow multiple top-levels (because the refactored wasi spec will be spread among many files) and that will change some things about the way `use` works - it is currently a bit fast and loose.